rust/yarp/build.rs

```diff
@@ -0,0 +1,507 @@
+use serde::Deserialize;
+use std::fs::File;
+use std::io::Write;
+use std::path::PathBuf;
+
+#[derive(Debug, Deserialize)]
+enum NodeParamType {
+    #[serde(rename = "node")]
+    Node,
+
+    #[serde(rename = "node?")]
+    OptionalNode,
+
+    #[serde(rename = "node[]")]
+    NodeList,
+
+    #[serde(rename = "string")]
+    String,
+
+    #[serde(rename = "constant")]
+    Constant,
+
+    #[serde(rename = "constant[]")]
+    ConstantList,
+
+    #[serde(rename = "location")]
+    Location,
+
+    #[serde(rename = "location?")]
+    OptionalLocation,
+
+    #[serde(rename = "location[]")]
+    LocationList,
+
+    #[serde(rename = "uint32")]
+    UInt32,
+
+    #[serde(rename = "flags")]
+    Flags
+}
+
+#[derive(Debug, Deserialize)]
+struct NodeParam {
+    name: String,
+    r#type: NodeParamType,
+    kind: Option<String>
+}
+
+#[derive(Debug, Deserialize)]
+struct Node {
+    name: String,
+    child_nodes: Option<Vec<NodeParam>>,
+    comment: String
+}
+
+#[derive(Debug, Deserialize)]
+struct Config {
+    nodes: Vec<Node>
+}
+
+/// The main function for the build script. This will be run by Cargo when
+/// building the library.
+/// 
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let file = std::fs::File::open(config_path())?;
+    let config: Config = serde_yaml::from_reader(file)?;
+
+    write_bindings(config)?;
+    Ok(())
+}
+
+/// Gets the path to the config.yml file at `[root]/config.yml`.
+///
+fn config_path() -> PathBuf {
+    cargo_manifest_path()
+        .join("../../config.yml")
```

Comment:
for publishing this to crates.io, `config.yml` is likely going to have to exist inside the crate root.

Reply:
Yeah, this is going to be part of a bigger conversation - I'm not sure how we're going to publish `yarp-sys` because it references the shared library built by the makefile. I'd imagine once we get things stable we'll need to spend some time to figure out how to get all of this stuff published.

Reply:
For the one `-sys` crate I've published (to a private crate server) which depended on a local C/C++ project, I pulled in the files I needed by using `include` in the `Cargo.toml`:

```toml
include = [
  "build.rs",
  "src/**/*",
  "vendor/foo/CMakeLists.txt",
  "vendor/foo/ios.toolchain.cmake",
  "vendor/foo/include/foo/api.h",
  "vendor/foo/include/foo/bar.h",
  "vendor/foo/source/bar.cpp",
  # etc
]
```

That was a couple of years ago, so it's not super fresh, but IIRC, this did require some finagling with the source tree to make `cargo` happye.